File: CodingInterviewsPackSolutionsCPP/Middle of the Linked List.cpp
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* tortoise = head;
        ListNode* hare = head;

        while(hare && hare->next){
            tortoise = tortoise->next;
            hare = hare->next->next;
        }
        return tortoise;
    }
};

File: CodingInterviewsJavaSolutions/Middle of the Linked List.java
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode tortoise = head;
        ListNode hare = head;

        while (hare != null && hare.next != null) {
            tortoise = tortoise.next;
            hare = hare.next.next;
        }
        return tortoise;
    }
}
